api/app/src/external/commonwell/link.ts

```diff
@@ -0,0 +1,181 @@
+  personId: string,
+  linkId: string
+): Promise<Link | void> => {
+  try {
+    // - get the patient link -> CommonWell.getPatientLink()
+    const allPatientLinksToPerson = await commonWell.getPatientLink(metriportQueryMeta, linkId);
+
+    if (
+      allPatientLinksToPerson._embedded &&
+      allPatientLinksToPerson._embedded.patientLink?.length
+    ) {
+      const correctLink = allPatientLinksToPerson._embedded.patientLink.filter(
+        link => link._links?.self?.href === linkId
+      );
+
+      //      - if the link exists
+      if (correctLink.length && correctLink[0].assuranceLevel) {
+        const assuranceLevel = parseInt(correctLink[0].assuranceLevel);
+
+        //      - if is >= LOLA 2
+        if (assuranceLevel >= 2) {
+          const person = await commonWell.searchPersonByUri(metriportQueryMeta, personId);
+          const personLink = convertPersonToLink(person, linkId);
+
+          if (personLink) return personLink;
+        }
+      }
+    }
+  } catch (error) {
+    const msg = `Failure retrieving`;
+    console.log(`${msg} - link id:`, linkId);
+    console.log(msg, error);
+    throw new Error(msg);
+  }
+};
+
+export const getPersonsAtCommonwell = async (patientId: string): Promise<Link[]> => {
+  try {
+    const personsResp = await commonWell.searchPersonByPatientDemo(metriportQueryMeta, patientId);
+
+    if (
+      personsResp &&
+      personsResp._embedded &&
+      personsResp._embedded.person &&
+      personsResp._embedded.person.length
+    ) {
+      return commonwellToLinks(personsResp._embedded.person);
+    }
+
+    throw new Error("No persons received");
+  } catch (error) {
+    const msg = `Failure retrieving persons`;
+    console.log(`${msg} - patient id:`, patientId);
+    console.log(msg, error);
+    throw new Error(msg);
+  }
+};
+
+const commonwellToLinks = (persons: Person[]): Link[] => {
+  const links: Link[] = [];
+
+  persons.forEach(person => {
+    const personLink = convertPersonToLink(person);
+
+    if (personLink) links.push(personLink);
+  });
+
+  return links;
+};
+
+const convertPersonToLink = (person: Person, linkId?: string): Link | null => {
+  const personId = getId(person);
+
+  if (personId) {
+    const personLink: Link = {
+      ...(linkId ? { id: linkId } : undefined),
+      entityId: personId,
+      potential: true,
+      source: LinkSource.commonWell,
+      patient: {
+        id: personId,
+        firstName:
+          person.details?.name?.length && person.details?.name[0].given?.length
+            ? person.details.name[0].given[0]
+            : "",
+        lastName:
+          person.details?.name?.length && person.details?.name[0].family?.length
+            ? person.details.name[0].family[0]
+            : "",
+        dob: person.details?.birthDate ? person.details?.birthDate : "", // YYYY-MM-DD
+        address: {
+          addressLine1:
+            person.details?.address?.length && person.details?.address[0].line
```

Comment:
can do the `person.details?.address?.length` check once before creating the object, and load an address object with the contents of `person.details?.address[0]`

Reply:
The types dont really match to be able to do that. For example addressLine1 in our schema but its line in CW.

Also each type in CW can be undefined but in our schemas the cant be so i need to check each one if its true

Reply:
What I mean is something like this (haven't tested this code):

```
if (personId) {
    const address = person.details?.address?.length ? person.details?.address[0] : undefined;
    const personName = person.details?.name?.length ? person.details?.name[0] : undefined;
    const personLink: Link = {
      ...(linkId ? { id: linkId } : undefined),
      entityId: personId,
      potential: true,
      source: LinkSource.commonWell,
      patient: {
        id: personId,
        firstName:
          personName && personName.given?.length
            ? personName.given[0]
            : "",
        lastName:
          personName && personName.family?.length
            ? personName.family[0]
            : "",
        dob: person.details?.birthDate ? person.details?.birthDate : "", // YYYY-MM-DD
        address: {
          addressLine1: address && address.line
```

This way we're not doing the same check/indexing many times, and shortens the code a bit.